designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,462 @@
+  =======================
+  "results": [
+    { ... another record ... },
+    {
+      "level": "error",
+      "message": {
+        "text": "'a' is not defined."
+      },
+      "locations": [
+        {
+          "physicalLocation": {
+            "artifactLocation": {
+              "uri": "file:///C:/Users/yiweiding/Documents/projects/test/1.js",
+              "index": 0
+            },
+            "region": {
+              "startLine": 1,
+              "startColumn": 1
+            }
+          }
+        }
+      ],
+      "suppressions": [
+        {
+          "kind": "inSource",
+          "justification": "Sample justification message."
+        }
+      ],
+      "ruleId": "no-undef",
+      "ruleIndex": 0
+    }
+  ]
+}
+```
+
+Input:
+
+```
+.\node_modules\.bin\eslint.cmd 1.js -f json -o 1.json
+```
+
+Output:
+
+```json
+JSON format
+[
+  {
+    "filePath":"C:\\Users\\yiweiding\\Documents\\projects\\test\\1.js",
+    "messages":[
+      { ... another record ... }
+    ],
+    "suppressedMessages":[
+      {
+        "ruleId":"no-undef",
+        "severity":2,
+        "message":"'a' is not defined.",
+        "line":1,
+        "column":1,
+        "nodeType":"Identifier",
+        "suppressions":[
+          {
+            "kind":"directive",
+            "justification":"Sample justification message."
+          }
+        ],
+        "messageId":"undef",
+        "endLine":1,
+        "endColumn":2
+      }
+    ],
+    "errorCount":0,
+    "warningCount":0,
+    "fixableErrorCount":0,
+    "fixableWarningCount":0,
+    "usedDeprecatedRules":[]
+  }
+]
+```
+
+### Parse description in comments as `Justification`
+
+We propose not to ignore the text after two or more dashes in comments and considered it as the justification of a suppression. The modification should be in `getDirectiveComments`.
+
+### Add `SuppressedLintMessage` inheriting from `LintMessage`
+
+According to this [comment](https://github.com/eslint/rfcs/pull/82#discussion_r733584774), we are going to have a `suppressedMessages` distinguished from `messages` in the output of ESLint. Items in `suppressedMessages` should have the property `suppressions`, a list of suppressions (`{kind: "directive", justification: "Justification message example."}`).
+
+### Reserve directive suppression
+
+In the step of apply directive suppression, ESLint will not remove the problems that should be disabled by directive comments. These problems will be reserved and added `{kind: "directive", justification: "Justification message example."}`.
+
+NOTE that the kind `directive` should be converted to `inSource` for SARIF according to [kind property](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317735).
+
+### Return `SuppressedLintMessage` in `Linter#verify` and `Linter#verifyAndFix`
+
+Since we have a new property `suppressedMessages` in `LintResult`, the output of `Linter#verify` and `Linter#verifyAndFix` will also have `suppressedMessages`.
```

Comment:
Return type of `Linter#verify`  is [`LintMessage[]`](https://github.com/eslint/eslint/blob/abe740ce68dcc9e5413df93b3d80a2e3260f1c18/lib/linter/linter.js#L1269). Is the plan to add `suppressedMessages` as a property on the array?

Reply:
No, as discussed before, `SuppressedLintMessage` will be a new property of `LintResult`, which is distinguished from `LintMessage`. So I think returning `{messages: LintMessage[], suppressedMessages: SuppressedLintMessage[]}` from `Linter#verify` would fit more for the ESLint design.